src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java

```diff
@@ -587,41 +586,98 @@ private int unpackInt32(final byte[] buffer, final int offset) {
+            final ByteBuffer lastBlockBuffer = ByteBuffer.allocate(BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK.length);
+            readFully(channel, lastBlockBuffer);
+            if (Arrays.equals(lastBlockBuffer.array(), BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK)) {
                 return FileTermination.HAS_TERMINATOR_BLOCK;
             }
-            final int bufsize = (int)Math.min(fileSize, BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE);
-            buf = new byte[bufsize];
-            raFile.seek(fileSize - bufsize);
-            raFile.read(buf);
-            for (int i = buf.length - BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK.length;
-                    i >= 0; --i) {
+
+            //if the last block isn't an empty gzip block, check to see if it is a healthy compressed block or if it's corrupted
+            final int bufsize = (int) Math.min(fileSize, BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE);
+            final byte[] bufferArray = new byte[bufsize];
+            channel.position(fileSize - bufsize);
+            readFully(channel, ByteBuffer.wrap(bufferArray));
+            for (int i = bufferArray.length - BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK.length;
+                 i >= 0; --i) {
                 if (!preambleEqual(BlockCompressedStreamConstants.GZIP_BLOCK_PREAMBLE,
-                        buf, i, BlockCompressedStreamConstants.GZIP_BLOCK_PREAMBLE.length)) {
+                                   bufferArray, i, BlockCompressedStreamConstants.GZIP_BLOCK_PREAMBLE.length)) {
                     continue;
                 }
-                final ByteBuffer byteBuffer = ByteBuffer.wrap(buf, i + BlockCompressedStreamConstants.GZIP_BLOCK_PREAMBLE.length, 4);
+                final ByteBuffer byteBuffer = ByteBuffer.wrap(bufferArray,
+                                                              i + BlockCompressedStreamConstants.GZIP_BLOCK_PREAMBLE.length,
+                                                              4);
                 byteBuffer.order(ByteOrder.LITTLE_ENDIAN);
-                final int totalBlockSizeMinusOne =  byteBuffer.getShort() & 0xFFFF;
-                if (buf.length - i == totalBlockSizeMinusOne + 1) {
+                final int totalBlockSizeMinusOne = byteBuffer.getShort() & 0xFFFF;
+                if (bufferArray.length - i == totalBlockSizeMinusOne + 1) {
                     return FileTermination.HAS_HEALTHY_LAST_BLOCK;
                 } else {
                     return FileTermination.DEFECTIVE;
                 }
             }
             return FileTermination.DEFECTIVE;
+        } catch (final Throwable e) {
+            exceptionThrown = true;
+            throw e;
         } finally {
-            raFile.close();
+            //if an exception was thrown we don't want to reset the position because that would be likely to throw again
+            //and suppress the initial exception
+            if(!exceptionThrown) {
```

Comment:
This code is not consistent with its use of spaces around `if` elements. Here there's no space between `if` and `(`. On line 679 there is a space there, but not after the `)`. On line 654 there are spaces both before and after.

The code should be consistent; putting a space before and after the `()`s is the java standard and is what most of the existing code does, so I would recommend formatting it that way.

The `try` on line 606 is also formatted inconsistently with respect to the surrounding code.